Goals system

Fesaa opened the pull request:
Sorry for the massive PR again :pensive: 

This PR adds the goal system, which adds a unified way to tell characters how to behave. And allowing them to share stuff. 

This also added a AfterSkill/Basic/ult BattleEvent, due to changes in how it was made. We needed somewhere, but I can't remember where.

I've also moved a few more things from the Battle Loop to powers, the correct chars. Only Yunli stuff remains, but she's such an exception that I don't think we can easily resolve it now. 


### Changes

DIFF 500 AV: https://www.diffchecker.com/HMTNIU18/ (new more dmg, Feixiao ults one more time)
![image](https://github.com/user-attachments/assets/9fbab03f-5886-4a40-a007-2daf157ae7d4)

DIFF 550 AV: https://www.diffchecker.com/jNldM8Jf/ (no changes)


I've been playing around with adding some different goals, and it's really fun and easy how you can change behaviour! 
I'm probably forgetting to say some stuff, but :v:

Fesaa committed:
Migrate decision-making to Goals system

Yunli needs fixing, rest is working the same or better. Hopefully smaller better commit after this, sorry about that.

Also still need to merge a bunch of goals that are very close to each other

Fesaa committed:
Fix too many techniques being passed to TriggerTechnique Loggable

Fesaa committed:
Add back sparkle talent power

This time without forcing it in the skill method

Fesaa committed:
Fix Yunli ult logic

- Ults if the next entity is an enemy
- Ults if only one more entity will have a turn, as we need one to trigger ult counter

Fesaa committed:
Log why decisions are made

Fesaa committed:
Add some TODO's, change stuff to streams, and remove energy check. Character should do this

Fesaa committed:
Small changes

- Fix a missing generic
- UltGoalResult will always be DO, just use ULT

Fesaa committed:
Save full battle logs when using my sanity check

Fesaa committed:
Add export dir to gitignore, also save last line there

Fesaa committed:
Move Concerto logic to its own class

Ideally we'd have an IBattle#addUnit IBattle#removeUnit instead of playing with the actionValue map ourselves.

https://www.diffchecker.com/UwKMssC2/

Fesaa committed:
Fix HuoHuo logic

Fesaa committed:
Fix Tingyun logic

Disable tryUlt at start for everyone, chars can enable themselves if they want ofc.

Fesaa committed:
Fix Yunli

Fesaa committed:
Unify some goals

Fesaa committed:
Fix package names

Fesaa committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into goals-system

# Conflicts:
#	StarRailBattleSim/src/battleLogic/Battle.java
#	StarRailBattleSim/src/battleLogic/log/lines/metrics/FinalDmgMetrics.java
#	StarRailBattleSim/src/battleLogic/log/lines/metrics/PostCombatPlayerMetrics.java

Fesaa committed:
Fix FeixiaoBroynaTurnGoal

Fesaa committed:
Use 550 AV, clean up code a little

Darkglade1 commented on StarRailBattleSim/src/battleLogic/Battle.java:
```diff
@@ -102,9 +110,35 @@ public AbstractEntity getNextUnit() {
     @Override
     public boolean isAboutToEnd() {
-        AbstractEntity next = findLowestAVUnit(actionValueMap);
+        AbstractEntity next = findLowestAVUnit();
```
Since you made a new (and better) method for getting the next unit to act (the getUnit(int index) method) we could just use getUnit(0) here instead of findLowestAVUnit() and probably delete the findLowestAVUnit() function.

Fesaa replied:
Good call, seems like my method handled it a little differently. So I fixed that as well, they behave the same now

Darkglade1 replied:
Also minor nitpick but I think `getNextUnit` would be a more better/descriptive name instead of `getUnit`

Darkglade1 commented on StarRailBattleSim/src/characters/AbstractCharacter.java:
```diff
@@ -95,41 +104,116 @@ public AbstractCharacter(String name, int baseHP, int baseAtk, int baseDef, int
+        super.takeTurn();
+        numTurnsMetric++;
+
+        //tryUltimate();
```
what is this commented out code here for?

Fesaa replied:
I forgot to delete, adding it there increases dmg for most teams (but not all, would need to look at log to write the correct goals so fix that). Was something to discuss, I commented it out to ensure values stay the same in this PR

Darkglade1 replied:
ah right, this would've been the tryUlt at start of turn event. In that case you can leave it commented it out for now, maybe also add a TODO to look into why this event would change the performance of many teams.

Darkglade1 commented on StarRailBattleSim/src/characters/AbstractCharacter.java:
```diff
@@ -95,41 +104,116 @@ public AbstractCharacter(String name, int baseHP, int baseAtk, int baseDef, int
-
         this.emit(BattleEvents::onUseUltimate);
+        this.useUltimate();
+        this.emit(BattleEvents::afterUseSkill);
```
should be afterUseUltimate

Darkglade1 commented on StarRailBattleSim/src/characters/feixiao/Feixiao.java:
```diff
@@ -45,6 +50,20 @@ public Feixiao() {
                 .setStat(PowerStat.CRIT_CHANCE, 12)
                 .setStat(PowerStat.DEF_PERCENT, 12.5f));
+
+        this.registerGoal(0, new FeixiaoBroynaTurnGoal(this));
```
Broyna -> Bronya

Darkglade1 commented on StarRailBattleSim/src/characters/lingsha/LingshaTurnGoal.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public TurnGoalResult determineAction() {
+        if (getBattle().getSkillPoints() >= 4) {
```
could probably make a separate TurnGoal called `UseExcessSkillPointsTurnGoal` or something and use that instead of having this in a character specific turn goal since a number of characters also have this behavior of skilling when there are spare skill points (pela/topaz/lingsha).

Darkglade1 commented on StarRailBattleSim/src/characters/moze/Moze.java:
```diff
@@ -39,10 +42,13 @@ public Moze() {
 
         preyPower = new MozePreyPower();
+
+        this.registerGoal(0, DontUltMissingPowerGoal.robin(this));
```
he also needs the UltAtEndOfBattle ult goal

Darkglade1 commented on StarRailBattleSim/src/characters/pela/Pela.java:
```diff
@@ -65,17 +66,17 @@ public void useUltimate() {
     }
 
-    public void takeTurn() {
+    /*public void takeTurn() {
```
delete commented out code?

Darkglade1 commented on StarRailBattleSim/src/characters/pela/PelaTurnGoal.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public TurnGoalResult determineAction() {
+        if (getBattle().getSkillPoints() > 0 && this.character.firstMove) {
```
might be good to make a `SkillFirstTurnGoal` and use that instead of having a character specific turn goal (March would also be able to use this goal).

Darkglade1 commented on StarRailBattleSim/src/characters/pela/PelaTurnGoal.java:
```diff
@@ -0,0 +1,21 @@
+        if (getBattle().getSkillPoints() > 0 && this.character.firstMove) {
+            this.character.firstMove = false;
+            return TurnGoalResult.SKILL;
+        } else if (getBattle().getSkillPoints() >= 4) {
```
`useExcessSkillPointsTurnGoal` instead if you decide to make it

Darkglade1 commented on StarRailBattleSim/src/characters/topaz/TopazTurnGoal.java:
```diff
@@ -0,0 +1,27 @@
+            return TurnGoalResult.BASIC;
+        }
+
+        if (character.firstMove) {
```
could probably break this out into `SkillFirstTurnGoal` and `UseExcessSkillPointsTurnGoal`, and then the only remaining logic in this goal would be using basic when Numby is enhanced.

Darkglade1 commented on StarRailBattleSim/src/characters/yunli/YunliEmergancyUlt.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    @Override
+    public UltGoalResult determineAction() {
```
is this goal meant to let yunli spend excess energy in case the battle is about to end? if so, bit confused as to what the purpose of checking the AV of the 2nd next character is for.

Fesaa replied:
As I understood her ult, she needs someone to take a turn, so if she ults at the very end. It won't do anything, this way we ult before the last turn is taken (regardless of enemy or not) and get an extra ult in.

I don't have her, so can't really check if this is how it works; but that was my thought process here

Darkglade1 replied:
Ah yes, that is how Yunli works. However the getBattle().isAboutToEnd() method checks if the next character to move is outside the battle's remaining AV, which means checking the 2nd next person to move won't do anything since currently that logic would only trigger when no more turns are remaining. 

I think what this goal should do instead is check that `getUnit(0)` is inside the battle's remaining AV, but that `getUnit(1)` is outside the battle's remaining AV, and then return DO in that case. And then you can just get rid of the `isAboutToEnd` check since it's not very useful for Yunli's case.

Fesaa replied:
Made the changes, didn't change any dmg values at 550 AV. Does on 500 AV

Darkglade1 commented on StarRailBattleSim/src/characters/yunli/YunliTurnGoal.java:
```diff
@@ -0,0 +1,27 @@
+            return TurnGoalResult.BASIC;
+        }
+
+        if (character.firstMove) {
```
could probably break this out into a `SkillFirstTurnGoal` and a `AlwaysSkillTurnGoal` with a threshold of 1.

Darkglade1 commented on StarRailBattleSim/src/characters/AbstractCharacter.java:
```diff
@@ -95,41 +104,116 @@ public AbstractCharacter(String name, int baseHP, int baseAtk, int baseDef, int
     }
 
-    public void useSkill() {
+    public void registerGoal(int priority, UltGoal<C> ultGoal) {
```
what about removing goals from a character's default set of goals? should probably have at least a `clearUltGoals` and `clearTurnGoals` functions in case a character needs drastically different goals for a certain team comp.

Fesaa replied:
Good point, I'd want to be able to remove a specific goal as well. Am going to need to think about how best to do this for a bit

Darkglade1 commented:
I see a lot of the changes was just adding <?> to some files so I just skimmed over the ones that were like that 😝 

"AfterSkill/Basic/ult BattleEvent," we thought we'd need them for certain relic sets bonus effects (4PC Ice and 4PC Fire), although apparently Star Rail's terrible wording kicks in again and apparently some of those set bonuses still actually occur before the ultimate is used? I'd have to actually check in game to confirm how they work, but at least these hooks will (probably) be useful in the future.

Stuff pretty much looks good 👍 just had some comments about factoring out some character specific turn goals into using a combination of general goals instead.

Fesaa committed:
Add pair, so we don't need javafx

Fesaa committed:
Use IBattle#getUnit for finding the next unit

Reworks the function to be more correct, and easier to read.

Fesaa committed:
Address easier comments

Fesaa committed:
Unify some more goals

Fesaa committed:
Actually fix getUnit

I hate floats

Fesaa committed:
Fix YunliEmergancyUlt

Fesaa committed:
Add TODO, add clear goals

Leaving it at clear all at the moment. Would like to add a clear specific one, but not finding a happy way to do ti

Fesaa committed:
Wrong file lol

Add TODO, add clear goals

Leaving it at clear all at the moment. Would like to add a clear specific one, but not finding a happy way to do ti

Fesaa committed:
Merge branch 'refs/heads/main' into goals-system

# Conflicts:
#	StarRailBattleSim/src/lightcones/hunt/SleepLikeTheDead.java
#	StarRailBattleSim/src/teams/PlayerTeam.java

Fesaa committed:
Fix merge errors

Fesaa committed:
Rename some fields and methods

Fesaa committed:
Move power removal to AbstractPower
https://www.diffchecker.com/KsrUSD9d/

Fesaa committed:
Fix metric logger not logging PreCombatPlayerMetrics
